### PR TITLE
Format `target: annotation = value?` expressions

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR implements formatting of `target: annotation (= value)?` expressions. 

It resulted in me going down a rabbit hole because I then noticed that tuples inside of subscript should not be parenthesized. So I fixed that too. 

## Test Plan

I added new tests. I fixed the `skip_magic_comma` test to run it with magic comma disabled (This test got me really confused because one test wanted to have trailing comma and this test wanted **not** to have trailing commas).

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5596** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5596" target="_blank">Graphite</a> 
    * **PR #5608** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5608" target="_blank">Graphite</a> 
      * **PR #5609** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5609" target="_blank">Graphite</a> 
        * **PR #5643** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5643" target="_blank">Graphite</a> 
          * **PR #5644** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5644" target="_blank">Graphite</a> 
            * **PR #5661** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5661" target="_blank">Graphite</a>  👈
        * **PR #5615** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5615" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5661?utm_source=stack-comment).

File: crates/ruff_python_formatter/src/statement/stmt_function_def.rs [hunk 1]
```diff
@@ -97,9 +97,9 @@ impl FormatRule<AnyFunctionDefinition<'_>, PyFormatContext<'_>> for FormatAnyFun
                     space(),
                     text("->"),
                     space(),
-                    return_annotation
-                        .format()
-                        .with_options(Parenthesize::IfBreaks)
+                    optional_parentheses(
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_function_def.rs:100]: Function defs are weird. They add parentheses around subscript expressions.

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@miscellaneous__force_pyi.py.snap [hunk 2]
```diff
@@ -89,29 +89,28 @@ def eggs() -> Union[str, int]: ...
...
-+
+ 
+-class F(A, C): ...
+ 
+-def spam() -> None: ...
 +class F(A, C):
 +    ...
 +
 +
 +def spam() -> None:
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/snapshots/black_compatibility@miscellaneous__force_pyi.py.snap:100]: Hmm, we'll need into single line bodies at some point. Potentially something that our suite refactor would simplify but we shouldn't be blocked by it.l

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__one_element_subscript.py.snap
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__one_element_subscript.py.snap:1]: :partying_face:

**Review by konstin — APPROVED** [review]

**Pull request merged**